File: src/frontend/src/lib/analytics.ts
import type { EntryWithDate, ReflectionWithDate } from "../backend.d";

export type RiskLevel = "Low" | "Medium" | "High";

export interface DailyRiskScore {
  level: RiskLevel;
  score: number;
  reasons: string[];
  suggestion: string;
}

export interface WeeklyAnalysis {
  habitCompletionPct: number;
  avgEnergyLevel: number;
  avgSleepHours: number;
  distractionFrequency: Record<string, number>;
  topDistraction: string | null;
  triggerChains: string[];
  suggestions: string[];
  daysAnalyzed: number;
}

export interface MomentumPoint {
  label: string;
  completionPct: number;
  avgScore: number;
}

function getLast7Days(entries: EntryWithDate[]): EntryWithDate[] {
  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - 7);
  const cutoffStr = cutoff.toISOString().slice(0, 10);
  return entries.filter((e) => e.date >= cutoffStr);
}

function getLast3DaysReflections(
  reflections: ReflectionWithDate[],
): ReflectionWithDate[] {
  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - 3);
  const cutoffStr = cutoff.toISOString().slice(0, 10);
  return reflections.filter((r) => r.date >= cutoffStr);
}

export function computeDailyRiskScore(
  entries: EntryWithDate[],
  reflections: ReflectionWithDate[],
): DailyRiskScore {
  let risk = 0;
  const reasons: string[] = [];

  const sorted = [...entries].sort((a, b) => b.date.localeCompare(a.date));
  const last = sorted[0];

  if (last) {
    const score = Number(last.entry.final_score);
    if (score < 5) {
      risk += 25;
      reasons.push(`Last score was ${score} (very low)`);
    } else if (score < 8 && !last.entry.deep_work_done) {
      risk += 20;
      reasons.push("Last entry: below threshold and no deep work");
    }
    if (Number(last.entry.screen_time) > 420) {
      risk += 15;
      reasons.push("High screen time yesterday (>7h)");
    }
  }

  // Check streak
  const streakBroken =
    sorted.length > 0 && Number(sorted[0].entry.final_score) < 8;
  if (streakBroken) {
    risk += 10;
    reasons.push("Recent streak failure");
  }

  // Check sleep from last 3 days reflections
  const recent3 = getLast3DaysReflections(reflections).filter(
    (r) => r.data.sleep_hours > 0,
  );
  if (recent3.length > 0) {
    const avgSleep =
      recent3.reduce((s, r) => s + r.data.sleep_hours, 0) / recent3.length;
    if (avgSleep < 6) {
      risk += 25;
      reasons.push(`Low avg sleep (${avgSleep.toFixed(1)}h over recent days)`);
    }
  }

  // Check energy level
  const recent3Energy = getLast3DaysReflections(reflections).filter(
    (r) => Number(r.data.energy_level) > 0,
  );
  if (recent3Energy.length > 0) {
    const avgEnergy =
      recent3Energy.reduce((s, r) => s + Number(r.data.energy_level), 0) /
      recent3Energy.length;
    if (avgEnergy <= 3) {
      risk += 20;
      reasons.push(`Low avg energy (${avgEnergy.toFixed(1)}/10)`);
    }
  }

  const capped = Math.min(risk, 100);
  const level: RiskLevel =
    capped >= 60 ? "High" : capped >= 30 ? "Medium" : "Low";

  let suggestion = "Maintain your current habits and stay disciplined.";
  if (level === "High") {
    suggestion =
      "Start deep work early before energy drops. Remove distractions now.";
  } else if (level === "Medium") {
    suggestion = "Be intentional today. Front-load your hardest tasks.";
  }

  return { level, score: capped, reasons, suggestion };
}

export function computeWeeklyAnalysis(
  entries: EntryWithDate[],
  reflections: ReflectionWithDate[],
): WeeklyAnalysis {
  const week = getLast7Days(entries);
  const daysAnalyzed = week.length;

  // Habit completion %
  let totalPossible = 0;
  let totalDone = 0;
  for (const e of week) {
    for (const [, done] of e.entry.tasks) {
      totalPossible++;
      if (done) totalDone++;
    }
  }
  const habitCompletionPct =
    totalPossible > 0 ? Math.round((totalDone / totalPossible) * 100) : 0;

  // Avg energy & sleep from reflections in last 7 days
  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - 7);
  const cutoffStr = cutoff.toISOString().slice(0, 10);
  const weekReflections = reflections.filter((r) => r.date >= cutoffStr);

  const energyRefs = weekReflections.filter(
    (r) => Number(r.data.energy_level) > 0,
  );
  const avgEnergyLevel =
    energyRefs.length > 0
      ? energyRefs.reduce((s, r) => s + Number(r.data.energy_level), 0) /
        energyRefs.length
      : 0;

  const sleepRefs = weekReflections.filter((r) => r.data.sleep_hours > 0);
  const avgSleepHours =
    sleepRefs.length > 0
      ? sleepRefs.reduce((s, r) => s + r.data.sleep_hours, 0) / sleepRefs.length
      : 0;

  // Distraction frequency
  const distractionFrequency: Record<string, number> = {};
  for (const ref of weekReflections) {
    for (const tag of ref.data.distraction_tags) {
      distractionFrequency[tag] = (distractionFrequency[tag] ?? 0) + 1;
    }
  }
  const topDistraction =
    Object.keys(distractionFrequency).sort(
      (a, b) => distractionFrequency[b] - distractionFrequency[a],
    )[0] ?? null;

  const triggerChains = detectTriggerChains(entries, reflections);

  const suggestions: string[] = [];
  if (habitCompletionPct < 50) {
    suggestions.push(
      "Habit completion is below 50%. Focus on fewer, higher-impact tasks first.",
    );
  }
  if (avgSleepHours > 0 && avgSleepHours < 6) {
    suggestions.push(
      "Sleep is critically low. Prioritize sleep to stabilize energy and focus.",
    );
  }
  if (avgEnergyLevel > 0 && avgEnergyLevel <= 3) {
    suggestions.push(
      "Energy is consistently low. Check sleep, nutrition, and morning routine.",
    );
  }
  if (topDistraction) {
    suggestions.push(
      `Most common distraction: ${topDistraction}. Create a dedicated removal plan.`,
    );
  }
  if (suggestions.length === 0) {
    suggestions.push("Keep the current momentum. Focus on consistency.");
  }

  return {
    habitCompletionPct,
    avgEnergyLevel,
    avgSleepHours,
    distractionFrequency,
    topDistraction,
    triggerChains,
    suggestions,
    daysAnalyzed,
  };
}

export function detectTriggerChains(
  entries: EntryWithDate[],
  reflections: ReflectionWithDate[],
): string[] {
  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - 7);
  const cutoffStr = cutoff.toISOString().slice(0, 10);

  const weekEntries = entries.filter((e) => e.date >= cutoffStr);
  const weekRefs = reflections.filter((r) => r.date >= cutoffStr);

  const refMap: Record<string, ReflectionWithDate> = {};
  for (const r of weekRefs) refMap[r.date] = r;

  let chain1Count = 0; // Low Sleep + Low Energy + Missed Deep Work
  let chain2Count = 0; // High Screen + Low Productivity
  let chain3Count = 0; // Low Energy + Phone/Social

  for (const e of weekEntries) {
    const ref = refMap[e.date];
    const deepWorkDone = e.entry.deep_work_done;
    const screenTime = Number(e.entry.screen_time);
    const ratio = e.entry.ratio;

    if (ref) {
      const sleep = ref.data.sleep_hours;
      const energy = Number(ref.data.energy_level);
      const tags = ref.data.distraction_tags;

      if (
        sleep > 0 &&
        sleep < 6 &&
        energy > 0 &&
        energy <= 3 &&
        !deepWorkDone
      ) {
        chain1Count++;
      }
      if (
        energy > 0 &&
        energy <= 3 &&
        (tags.includes("Phone distraction") || tags.includes("Social media"))
      ) {
        chain3Count++;
      }
    }

    if (screenTime > 420 && ratio < 0.4) {
      chain2Count++;
    }
  }

  const chains: string[] = [];
  if (chain1Count >= 2) {
    chains.push("Low Sleep → Low Energy → Missed Deep Work");
  }
  if (chain2Count >= 2) {
    chains.push("High Screen Time → Low Productivity");
  }
  if (chain3Count >= 2) {
    chains.push("Low Energy → Phone Distraction → Missed Work");
  }

  return chains;
}

export function computeMomentumSeries(
  entries: EntryWithDate[],
): MomentumPoint[] {
  const points: MomentumPoint[] = [];

  for (let week = 3; week >= 0; week--) {
    const end = new Date();
    end.setDate(end.getDate() - week * 7);
    const start = new Date(end);
    start.setDate(start.getDate() - 6);

    const startStr = start.toISOString().slice(0, 10);
    const endStr = end.toISOString().slice(0, 10);

    const weekEntries = entries.filter(
      (e) => e.date >= startStr && e.date <= endStr,
    );

    let totalPossible = 0;
    let totalDone = 0;
    let totalScore = 0;

    for (const e of weekEntries) {
      for (const [, done] of e.entry.tasks) {
        totalPossible++;
        if (done) totalDone++;
      }
      totalScore += Number(e.entry.final_score);
    }

    const completionPct =
      totalPossible > 0 ? Math.round((totalDone / totalPossible) * 100) : 0;
    const avgScore =
      weekEntries.length > 0
        ? Math.round((totalScore / weekEntries.length) * 10) / 10
        : 0;

    const label = week === 0 ? "This week" : `${week}w ago`;
    points.push({ label, completionPct, avgScore });
  }

  return points;
}

File: local-export/backend/analytics.py
"""Rule-based behavioural analysis engine.

All logic is deterministic — no external AI dependencies.
The same rules are used by both the weekly analysis and the daily risk scorer.
"""

from __future__ import annotations

from collections import Counter
from typing import Any, Dict, List, Optional


# ── internal helpers ───────────────────────────────────────────────────────────

def _avg(values: List[float]) -> float:
    """Return mean of a non-empty list, else 0.0."""
    return sum(values) / len(values) if values else 0.0


def _top_tag(tags_lists: List[List[str]]) -> Optional[str]:
    """Return the most frequently occurring tag across all lists."""
    counter: Counter = Counter()
    for tags in tags_lists:
        counter.update(tags)
    if not counter:
        return None
    return counter.most_common(1)[0][0]


# ── weekly analysis ────────────────────────────────────────────────────────────

def compute_weekly_analysis(
    entries: List[Dict[str, Any]],
    reflections: List[Dict[str, Any]],
) -> Dict[str, Any]:
    """Analyse the last 7 days of entries + reflections.

    Parameters
    ----------
    entries:
        List of entry dicts (filtered to last 7 days, most-recent first).
    reflections:
        Matching reflection dicts for those days.

    Returns
    -------
    dict matching the WeeklyAnalysis schema.
    """
    days = len(entries)

    # ── habit completion ───────────────────────────────────────────────────────
    total_possible = 0
    total_done = 0
    for entry in entries:
        tasks = entry.get("tasks", [])
        for item in tasks:
            # tasks stored as [[key, bool], ...] or [(key, bool), ...]
            if isinstance(item, (list, tuple)) and len(item) == 2:
                total_possible += 1
                if item[1]:
                    total_done += 1
    habit_completion_pct = (total_done / total_possible * 100) if total_possible else 0.0

    # ── average energy & sleep ─────────────────────────────────────────────────
    energy_values = [
        r["energy_level"]
        for r in reflections
        if r.get("energy_level", 0) > 0
    ]
    sleep_values = [
        r["sleep_hours"]
        for r in reflections
        if r.get("sleep_hours", 0.0) > 0.0
    ]
    avg_energy = _avg(energy_values)
    avg_sleep = _avg(sleep_values)

    # ── top distraction tag ────────────────────────────────────────────────────
    all_tag_lists = [r.get("distraction_tags", []) for r in reflections]
    top_distraction = _top_tag(all_tag_lists)

    # ── trigger chain detection ────────────────────────────────────────────────
    trigger_chains = _detect_trigger_chains(entries, reflections)

    # ── actionable suggestions ─────────────────────────────────────────────────
    suggestions = _generate_suggestions(
        habit_completion_pct=habit_completion_pct,
        avg_energy=avg_energy,
        avg_sleep=avg_sleep,
        top_distraction=top_distraction,
        trigger_chains=trigger_chains,
        entries=entries,
    )

    return {
        "habit_completion_pct": round(habit_completion_pct, 1),
        "avg_energy_level": round(avg_energy, 1),
        "avg_sleep_hours": round(avg_sleep, 1),
        "top_distraction": top_distraction,
        "trigger_chains": trigger_chains,
        "suggestions": suggestions,
        "days_analyzed": days,
    }


def _detect_trigger_chains(
    entries: List[Dict[str, Any]],
    reflections: List[Dict[str, Any]],
) -> List[Dict[str, Any]]:
    """Detect common behavioural cause-effect chains."""
    chains: List[Dict[str, Any]] = []

    ref_map: Dict[str, Dict[str, Any]] = {r["date"]: r for r in reflections}

    low_sleep_days = 0
    low_energy_days = 0
    missed_deep_work_days = 0
    high_screen_days = 0
    phone_distraction_days = 0

    for entry in entries:
        date = entry["date"]
        ref = ref_map.get(date, {})

        sleep = ref.get("sleep_hours", 0.0)
        energy = ref.get("energy_level", 0)
        screen = entry.get("screen_time", 0)
        deep_work = entry.get("deep_work_done", False)
        tags: List[str] = ref.get("distraction_tags", [])

        if 0 < sleep < 6:
            low_sleep_days += 1
        if 0 < energy <= 3:
            low_energy_days += 1
        if not deep_work:
            missed_deep_work_days += 1
        if screen > 420:
            high_screen_days += 1
        if "Phone distraction" in tags or "Social media" in tags:
            phone_distraction_days += 1

    n = max(len(entries), 1)

    # Pattern A: Low Sleep → Low Energy → Missed Deep Work
    if (
        low_sleep_days >= max(1, n // 3)
        and low_energy_days >= max(1, n // 3)
        and missed_deep_work_days >= max(1, n // 2)
    ):
        chains.append({
            "factors": ["Low Sleep", "Low Energy", "Missed Deep Work"],
            "description": (
                f"Detected {low_sleep_days}/{n} days with low sleep, "
                f"{low_energy_days}/{n} with low energy, and "
                f"{missed_deep_work_days}/{n} with missed deep work. "
                "Poor sleep directly impairs cognitive energy, making deep work unlikely."
            ),
        })

    # Pattern B: Low Energy → High Screen Time → Missed Deep Work
    if (
        low_energy_days >= max(1, n // 3)
        and high_screen_days >= max(1, n // 3)
        and missed_deep_work_days >= max(1, n // 2)
    ):
        chains.append({
            "factors": ["Low Energy", "High Screen Time", "Missed Deep Work"],
            "description": (
                f"Detected {low_energy_days}/{n} low-energy days coinciding with "
                f"{high_screen_days}/{n} high-screen-time days. "
                "Low energy tends to drive passive screen consumption instead of focused work."
            ),
        })

    # Pattern C: Phone / Social Media distraction → Missed Deep Work
    if (
        phone_distraction_days >= max(1, n // 3)
        and missed_deep_work_days >= max(1, n // 2)
    ):
        chains.append({
            "factors": ["Phone / Social Media Distraction", "Missed Deep Work"],
            "description": (
                f"Phone or social media tagged on {phone_distraction_days}/{n} days, "
                f"with deep work missed on {missed_deep_work_days}/{n} days. "
                "Device access during work hours is breaking focus sessions."
            ),
        })

    return chains


def _generate_suggestions(
    habit_completion_pct: float,
    avg_energy: float,
    avg_sleep: float,
    top_distraction: Optional[str],
    trigger_chains: List[Dict[str, Any]],
    entries: List[Dict[str, Any]],
) -> List[str]:
    suggestions: List[str] = []

    if avg_sleep > 0 and avg_sleep < 6:
        suggestions.append(
            f"Average sleep is {avg_sleep:.1f} h — aim for at least 7 h to restore cognitive energy."
        )

    if avg_energy > 0 and avg_energy < 4:
        suggestions.append(
            "Average energy is low. Schedule deep work in your highest-energy window (usually morning)."
        )

    if habit_completion_pct < 50:
        suggestions.append(
            f"Only {habit_completion_pct:.0f}% of habits completed this week. "
            "Reduce optional commitments and protect your core habits."
        )

    if top_distraction in ("Phone distraction", "Social media"):
        suggestions.append(
            "Phone/social media is your top distraction. "
            "Remove your phone from your workspace before starting deep work."
        )
    elif top_distraction == "Poor planning":
        suggestions.append(
            "Poor planning keeps recurring. Write tomorrow's top 3 tasks the night before."
        )
    elif top_distraction == "Procrastination":
        suggestions.append(
            "Procrastination is the top pattern. Use a 2-minute start rule: "
            "open the task and work for exactly 2 minutes before deciding to stop."
        )
    elif top_distraction == "Low energy":
        suggestions.append(
            "Low energy is the primary obstacle. Prioritise sleep, hydration, and "
            "schedule demanding tasks earlier in the day."
        )

    # Chain-based suggestions
    chain_factors = {f for c in trigger_chains for f in c["factors"]}
    if "Low Sleep" in chain_factors and "Missed Deep Work" in chain_factors:
        suggestions.append(
            "Sleep is directly correlated with missed deep work. "
            "Set a hard sleep deadline and treat it like a non-negotiable task."
        )

    # Check average screen time
    screen_values = [e["screen_time"] for e in entries if e.get("screen_time", 0) > 0]
    if screen_values and _avg(screen_values) > 360:
        suggestions.append(
            f"Average screen time is {_avg(screen_values):.0f} min/day. "
            "Install a screen-time limiter and set a 300-minute daily cap."
        )

    if not suggestions:
        suggestions.append(
            "Solid week. Keep your current routine and incrementally raise the bar."
        )

    return suggestions


# ── daily risk score ───────────────────────────────────────────────────────────

def compute_risk_score(
    latest_entry: Optional[Dict[str, Any]],
    recent_reflections: List[Dict[str, Any]],  # most-recent first, up to 3
    current_streak: int,
) -> Dict[str, Any]:
    """Compute a 0-100 risk score for today's likelihood of failure.

    Risk points:
    - last entry final_score < 5        → +25
    - last entry no deep work           → +20
    - avg sleep (last 3) < 6 h          → +25
    - avg energy (last 3) <= 3          → +20
    - last entry screen_time > 420 min  → +15
    - current_streak == 0               → +10
    """
    risk = 0
    reasons: List[str] = []
    suggestions: List[str] = []

    if latest_entry:
        if latest_entry.get("final_score", 10) < 5:
            risk += 25
            reasons.append(
                f"Last entry score was {latest_entry['final_score']} (below 5)."
            )
            suggestions.append(
                "Identify the specific task that failed yesterday and address it first today."
            )

        if not latest_entry.get("deep_work_done", True):
            risk += 20
            reasons.append("Deep work was not completed yesterday.")
            suggestions.append(
                "Block the first 2 hours of today exclusively for deep work before any screen time."
            )

        if latest_entry.get("screen_time", 0) > 420:
            risk += 15
            reasons.append(
                f"Screen time was {latest_entry['screen_time']} min yesterday (above 7 h)."
            )
            suggestions.append(
                "Cap screen time to 300 min today. Set a phone-away rule during work blocks."
            )

    # Sleep & energy from last 3 reflections
    sleep_vals = [
        r["sleep_hours"]
        for r in recent_reflections
        if r.get("sleep_hours", 0.0) > 0.0
    ]
    energy_vals = [
        r["energy_level"]
        for r in recent_reflections
        if r.get("energy_level", 0) > 0
    ]

    if sleep_vals:
        avg_sleep = _avg(sleep_vals)
        if avg_sleep < 6:
            risk += 25
            reasons.append(
                f"Average sleep over the last {len(sleep_vals)} night(s) is {avg_sleep:.1f} h."
            )
            suggestions.append("Go to sleep at least 7.5 hours before your alarm tonight.")

    if energy_vals:
        avg_energy = _avg(energy_vals)
        if avg_energy <= 3:
            risk += 20
            reasons.append(
                f"Average energy level over the last {len(energy_vals)} day(s) is {avg_energy:.1f}/10."
            )
            suggestions.append(
                "Start with a short physical warm-up before attempting deep work."
            )

    if current_streak == 0:
        risk += 10
        reasons.append("Current streak is 0 — no positive momentum.")
        suggestions.append(
            "Use today to start a new streak. One full session of deep work changes the trajectory."
        )

    # Clamp to 0-100
    risk = min(risk, 100)

    if risk >= 60:
        level = "High"
    elif risk >= 30:
        level = "Medium"
    else:
        level = "Low"

    if not reasons:
        reasons.append("No major risk factors detected based on recent data.")
    if not suggestions:
        suggestions.append(
            "Keep the current routine. Start deep work early to lock in the streak."
        )

    return {
        "score": risk,
        "level": level,
        "reasons": reasons,
        "suggestions": suggestions,
    }


# ── streak computation ─────────────────────────────────────────────────────────

def compute_streak(
    entries: List[Dict[str, Any]],  # ALL entries
    streak_threshold: int = 7,
) -> Dict[str, Any]:
    """Count current and longest streaks.

    A streak day requires:
    - final_score >= streak_threshold
    - deep_work_done == True
    """
    if not entries:
        return {"current_streak": 0, "longest_streak": 0, "last_reset_date": None}

    # Sort most-recent first
    sorted_entries = sorted(entries, key=lambda e: e["date"], reverse=True)

    # Current streak: count from the most recent entry backwards
    current = 0
    for entry in sorted_entries:
        qualifies = (
            entry.get("final_score", 0) >= streak_threshold
            and entry.get("deep_work_done", False)
        )
        if qualifies:
            current += 1
        else:
            break

    # Longest streak: scan in chronological order
    longest = 0
    run = 0
    last_reset: Optional[str] = None
    for entry in reversed(sorted_entries):
        qualifies = (
            entry.get("final_score", 0) >= streak_threshold
            and entry.get("deep_work_done", False)
        )
        if qualifies:
            run += 1
            longest = max(longest, run)
        else:
            if run > 0:
                last_reset = entry["date"]
            run = 0

    if current == 0 and sorted_entries:
        last_reset = sorted_entries[0]["date"]

    return {
        "current_streak": current,
        "longest_streak": longest,
        "last_reset_date": last_reset,
    }
